src/js/core/methods/EthereumSignTypedData.js

```diff
@@ -0,0 +1,226 @@
+import { validateParams, getFirmwareRange } from './helpers/paramsValidator';
+import { validatePath } from '../../utils/pathUtils';
+import { getEthereumNetwork } from '../../data/CoinInfo';
+import { toChecksumAddress, getNetworkLabel } from '../../utils/ethereumUtils';
+import type { CoreMessage, EthereumNetworkInfo } from '../../types';
+import type {
+    MessageResponse,
+    EthereumTypedDataSignature,
+    EthereumTypedDataStructAck,
+    EthereumTypedDataValueRequest,
+} from '../../types/trezor/protobuf';
+import { ERRORS } from '../../constants';
+import type { EthereumSignTypedData as EthereumSignTypedDataParams } from '../../types/networks/ethereum';
+import { getFieldType, parseArrayType, encodeData } from './helpers/ethereumTypeDataConversions';
+
+type Params = {
+    ...EthereumSignTypedDataParams,
+    path: number[],
+    network?: EthereumNetworkInfo,
+};
+
+// Sanitization is used for T1 as eth-sig-util does not support BigInt
+function sanitizeData(data) {
+  switch(Object.prototype.toString.call(data)) {
+    case '[object Object]':
+      let entries = Object.keys(data).map((k) => [k, sanitizeData(data[k])]);
+      return Object.fromEntries(entries);
+
+    case '[object Array]':
+      return data.map((v) => sanitizeData(v));
+
+    case '[object BigInt]':
+      return data.toString();
+
+    default:
+      return data;
+  }
+}
+
+export default class EthereumSignTypedData extends AbstractMethod {
+    params: Params;
+
+    constructor(message: CoreMessage) {
+        super(message);
+
+        this.requiredPermissions = ['read', 'write'];
+
+        const { payload } = message;
+
+        // validate incoming parameters
+        validateParams(payload, [
+            { name: 'path', obligatory: true },
+            { name: 'data', type: 'object', obligatory: true },
+            { name: 'metamaskV4Compatibility', type: 'boolean', obligatory: true },
+        ]);
+
+        const path = validatePath(payload.path, 3);
+        const network = getEthereumNetwork(path);
+        this.firmwareRange = getFirmwareRange(this.name, network, this.firmwareRange);
+
+        this.info = getNetworkLabel('Sign #NETWORK typed data', network);
+
+        const { data, metamaskV4Compatibility } = payload;
+
+        this.params = {
+            path,
+            network,
+            data,
+            metamaskV4Compatibility,
+        };
+    }
+
+    async run() {
+        const cmd = this.device.getCommands();
+        const { path: address_n, network, data, metamaskV4Compatibility } = this.params;
+
+        const { types, primaryType, domain, message } = sigUtil.TypedDataUtils.sanitizeData(data);
+
+        let response: MessageResponse<
+            | 'EthereumTypedDataStructRequest'
+            | 'EthereumTypedDataValueRequest'
+            | 'EthereumTypedDataSignature',
+        >;
+
+        if (this.device.features.model === '1') {
+          // For Model 1 we use EthereumSignTypedHash
```

Comment:
Hi,
Thank you for the comment.

In this case we can even create a different method here `EthereumSignTypedHash.js` and expect the user to give us the hashes for domain separator and message. This will look exactly like ledgers library with EIP-712.

However if we don't do the conversion implicitly we are exposing a difference in implementations of T1 and TT to the user. Something that in my opinion they should not worry about (or even know about). Every client who needs to use EIP-712 should implement two different ways for T1 and TT. But if you have strong non-discussable opinions about adding the dependency then doing as you suggested would be the only way.

Reply:
So, the final decision how we would like to have it done at the moment is:
- plugin approach, implementator will provide additional data needed for model 1. Implementator typically has the required dependency on his side installed already. We might decide later to support more ethereum features natively in trezor-connect but this is not the case now. 
- there will be only 1 method and it will accept params for both model 1 and model T. Params for model 1 will be required in case model 1 is connected. Using 2 different methods isn't very practical as implementator in some case might not know which device is connected so we don't want to force him to make decisions about which method should be called.

Reply:
This is now done in 0e0354e0f046fca44bbb14167372c29ddf7e48c3